fix #688: add csrf & encrypted page tokens for scans

groovecoder opened the pull request:
Note: This breaks the "Scan another email" form at the bottom of the `/scan` results. Not sure yet how to un-break it, but I need some other eyes on this ASAP.

pdehaan commented on views/partials/scan_form.hbs:
```diff
@@ -1,4 +1,6 @@
 <form action="/scan" class="email-scan form-group" method="post" data-no-csrf novalidate>
```
Personal curiosity, but now that we're passing hidden `_csrf` fields w `{{csrfToken}}`s, do we still need/want `data-no-csrf` form attributes?

psiinon replied:
Yeah, if you're supplying a CSRF token then you should definitely remove the data-no-csrf attribute :)

groovecoder replied:
Ah, I think I remember now ... this attribute is a signal to ZAP? Or another of our scanners?

pdehaan commented on controllers/scan.js:
```diff
@@ -1,15 +1,68 @@
+    }
+    const decryptedPageToken = _decryptPageToken(encryptedPageToken);
+    if (!_validatePageToken(decryptedPageToken, req)) {
+      throw new FluentError("error-invalid-page-token");
```
I didn't find `"error-no-page-token"` (L57) or `"error-invalid-page-token"` in any of the .ftl files. Not sure if we need to add the strings still since they appear to be user/bot facing.

groovecoder replied:
Yes, will need to add some strings for this. But I don't know if it needs @betsymi to write them, or if I can just write them myself. @betsymi - do you want to write error messages here?

betsymi replied:
Yes, happy to. I'm having a hard time following what the errors are themselves though. @groovecoder could you help me out by drafting what these are supposed to communicate, and I can work on making them more user-friendly?

groovecoder replied:
@betsymi @flodolo right now I just have a single error string: "Your network may be performing too many scans. Please try again later."

Is that good?

flodolo replied:
I wonder if that's a bit confusing for normal users? (i.e. what's "my network"?) Not sure what's the target and frequency of this message showing up though.

Maybe something like: "We detected too many scans originating from your address and the request has been blocked. Please try again later."

betsymi replied:
What makes this error appear? Is it that a bot is trying to scan too many email addresses in quick succession?

groovecoder replied:
So far yes - this is aimed at bots. The normal users who are likely to encounter this would be people who are using "Scan another email" box repeatedly.

I agree "network" may not be the right term. But "address" might feel weird too - will people think we're somehow looking at their physical address? Should we be afraid of using the term "IP Address"? Google and many others use it all the time when they don't want me using their services from Tor.

(But as a scraper-writing colleague of mine pointed out, we shouldn't be anti-bot 😜 ... i.e., we don't really want humans to be scanning  dozens of email addresses either.)

betsymi replied:
You tried to search too many emails in a short time period. For security reasons, this IP address is temporarily blocked from new searches. You can try again later.

groovecoder commented:
Now that it's the next day, I think the encrypted page token effectively provides CSRF protection. @psiinon - what do you think?

psiinon commented:
Quite possibly - how is the encrypted page token created, stored and used?

jvehent commented on controllers/home.js:
```diff
@@ -1,10 +1,30 @@
 
+
+function _generatePageToken(req) {
+  const pageToken = {ip: req.ip, date: new Date(), nonce: uuidv4()};
```
just to sanity check: `req.ip` contains the client ip from the x-forwarded-for, not the load balancer?

groovecoder replied:
Correct, because we've got [`app.enable("trust proxy");`](https://expressjs.com/en/4x/api.html#trust.proxy.options.table)

jvehent replied:
From that doc: `If true, the client’s IP address is understood as the left-most entry in the X-Forwarded-* header.`

Which is a big fail, because we know bots inject x-forwarded-for headers to pass random ips as the leftmost entry. @Micheletto can confirm that our glb doesn't sanitize those, which is sad, but effectively means this control won't work.

jvehent replied:
Ok, that library supports `Trusting the nth hop from the front-facing proxy server as the client.` Is that how we're currently setup?

g-k replied:
Doesn't look like it? I think this should be a number instead of `true` now: https://github.com/mozilla/blurts-server/blob/1e2bc90b3e20a0a59438c0f6b9bda9241b831ed8/server.js#L136-L140

Sanitizing XFF seems like pretty basic functionality, disappointing that GLB doesn't do that.

Micheletto replied:
> @Micheletto can confirm that our glb doesn't sanitize those, which is sad, but effectively means this control won't work.

That is correct.  We could set a header that is passed to the app with the real-ip though.

groovecoder replied:
Nope, but I can set it to whatever N we want. Do we know it will always be the *n*th hop from the front-facing proxy server?

groovecoder replied:
I'm going to drop the `trust proxy` setting here, and we're going to have nginx explicitly set a custom header `X-Real-IP` that I will read from.

jvehent commented on controllers/home.js:
```diff
@@ -1,10 +1,30 @@
+
+function _generatePageToken(req) {
+  const pageToken = {ip: req.ip, date: new Date(), nonce: uuidv4()};
+  const cipher = crypto.createCipher("aes-256-cbc", AppConstants.COOKIE_SECRET);
```
can we use `aes-256-gcm` instead?

g-k replied:
In that case we need to remember to update https://github.com/mozilla/blurts-server/pull/747/files#diff-3e18cca2cad1fa80c47052b459cd06a0R16 too

g-k replied:
Also, is the confidentiality of the token data important? Could we sign the data and submit nonce, date, and sig instead? I guess that might be larger.

groovecoder replied:
When I read [how to use GCM](https://lollyrock.com/posts/nodejs-encryption/), it looks like we would need a per-encryption IV, and we don't currently have a back-end data store for holding those IVs and associating them with each session ID. Is `-cbc` okay "for now" ?

jvehent replied:
The IV and AAD should just be stored in cleartext alongside the encrypted data, but yes, cbc is fine for now.

jvehent commented on controllers/home.js:
```diff
@@ -1,10 +1,30 @@
+function _generatePageToken(req) {
+  const pageToken = {ip: req.ip, date: new Date(), nonce: uuidv4()};
+  const cipher = crypto.createCipher("aes-256-cbc", AppConstants.COOKIE_SECRET);
+  const encryptedPageToken = [cipher.update(JSON.stringify(pageToken), "utf8", "hex"), cipher.final("hex")].join("");
```
nit: base64 may be shorter than hex

groovecoder commented:
tl;dr the code 

1. On a GET to `/` I create a token of the client IP, timestamp, and a nonce. I encrypt the token using our `COOKIE_SECRET` env var value.
2. The scan form on `/` includes the encrypted token in a `<input type="hidden">` field.
3. POST to `/scan` receives the token, decrypts it, checks that the request IP address matches the IP address in the token, and that the timestamp in the token is less than 60 seconds ago. (60s value configurable by `PAGE_TOKEN_TIME` env var so we can adjust if too many humans are getting errors.)

The protections this adds:

1. Every `POST /scan` needs a valid unique token that it receives from the form in `/` (effective CSRF protection, right?)
2. Each IP has a unique token, so a token can't be re-used across IP addresses (i.e., each client has to `GET /` to get its own token)
3. The token expires in 60s, so each client has to `GET /` at least once every 60s

g-k commented on package.json:
```diff
@@ -14,6 +14,7 @@
     "client-oauth2": "^4.2.1",
     "client-sessions": "^0.8.0",
     "cpr": "^3.0.1",
+    "csurf": "^1.9.0",
```
NB: This is kind of unmaintained (last published 3 years ago; older open PRs). https://npms.io/search?q=csurf

I think it's solid though and don't know of any alternatives of the top of my head.

groovecoder commented:
I added the same token to the "Scan another email" form to fix that form for humans that use the form before the page token expires. Also added an error message when the token is invalid, so @flodolo will have to review before merge.

flodolo commented:
> I added the same token to the "Scan another email" form to fix that form for humans that use the form before the page token expires. Also added an error message when the token is invalid, so @flodolo will have to review before merge.

Do we expect more strings, or that's just this one?

groovecoder commented:
Updated with new error message. @flodolo - how does this one look?
